logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,215 @@
+                    } else {
+                        throw new SecretStoreException.NotLogstashKeyStore(String.format("Found a file at %s, but it is not a valid logstash keystore.",
+                                keyStorePath.toAbsolutePath().toString()), ioe);
+                    }
+                }
+                byte[] marker = retrieveSecret(logstashMarker);
+                if (marker == null) {
+                    throw new SecretStoreException.NotLogstashKeyStore(String.format("Found a keystore at %s, but it is not a logstash keystore.",
+                            keyStorePath.toAbsolutePath().toString()));
+                }
+                LOGGER.debug("Using existing keystore at {}", keyStorePath.toAbsolutePath().toString());
+            } catch (NoSuchFileException noSuchFileException) {
+                LOGGER.info("Keystore not found at {}. Creating new keystore.", keyStorePath.toAbsolutePath().toString());
+                //create the keystore on disk with a default entry to identify this as a logstash keystore
+                try (final OutputStream os = Files.newOutputStream(keyStorePath)) {
+                    keyStore = KeyStore.Builder.newInstance(KEYSTORE_TYPE, null, protectionParameter).getKeyStore();
+                    SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+                    byte[] base64 = SecretStoreUtil.base64Encode(LOGSTASH_MARKER.getBytes(StandardCharsets.UTF_8));
+                    SecretKey secretKey = factory.generateSecret(new PBEKeySpec(SecretStoreUtil.asciiBytesToChar(base64)));
+                    keyStore.setEntry(logstashMarker.toExternalForm(), new KeyStore.SecretKeyEntry(secretKey), protectionParameter);
+                    keyStore.store(os, this.keyStorePass);
+
+                    PosixFileAttributeView attrs = Files.getFileAttributeView(keyStorePath, PosixFileAttributeView.class);
+                    if (attrs != null) {
+                        try {
+                            //wrapped in a system mostly for testing, but could be used for more restrictive defaults
+                            attrs.setPermissions(PosixFilePermissions.fromString(System.getProperty("logstash.keystore.java.file.perms", "rw-rw----")));
+                        } catch (Exception e) {
+                            try {
+                                Files.delete(keyStorePath);
+                            } catch (IOException ioe) {
+                                //do nothing
+                            }
+                            throw new SecretStoreException.CreateException("Error can not set file permissions on Java keystore", e);
+                        }
+                    }
+                    LOGGER.info("Keystore created at {}", keyStorePath.toAbsolutePath().toString());
+                } catch (Exception e) {
+                    throw new SecretStoreException.CreateException("Error can not create Java keystore", e);
```

Comment:
recommend: "Failed to create keystore"

"Error" in the current sentence seems like a noun and we are saying a thing called "Error" could not create something.

Do we need to specify "Java keystore" ? Can we say "Logstash key store" or "secret store" (whatever we are calling this)

Reply:
Thanks, my grammar is poor suggestions are always welcome. 

will upper case Logstash and keep a consistent 'keystore' reference. The idea is a secret store, the implementation is here is a keystore